File: my-folder/0543-diameter-of-binary-tree/solution.c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
int max(int a, int b) {
    return a > b ? a : b;
}

int maxDepth(struct TreeNode* root, int* diameter) {
    if (root == NULL)
        return 0;
    
    int left_depth = maxDepth(root->left, diameter);
    int right_depth = maxDepth(root->right, diameter);
    
    // Update diameter if needed
    *diameter = max(*diameter, left_depth + right_depth);
    
    // Return the depth of the current node
    return 1 + max(left_depth, right_depth);
}

int diameterOfBinaryTree(struct TreeNode* root) {
    if (root == NULL)
        return 0;
    
    int diameter = 0; // Initialize diameter
    
    maxDepth(root, &diameter);
    
    return diameter;
}

File: my-folder/0543-diameter-of-binary-tree/solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int diameterOfBinaryTree(TreeNode root) {
        int [] d = new int[1];
        dia(root,d);
        return d[0];
    }

    public int dia(TreeNode root,int [] maxi){
        if(root == null) return 0;

        int left = dia(root.left,maxi);
        int right = dia(root.right,maxi);

        maxi[0] = Math.max(left+right,maxi[0]);

        return 1+ Math.max(left,right);
    }
}
